src/sentry/discover/dataset_split.py

```diff
@@ -0,0 +1,384 @@
+    errors_builder = ErrorsQueryBuilder(
+        Dataset.Events,
+        params,
+        snuba_params=snuba_dataclass,
+        query=query,
+        selected_columns=selected_columns,
+        equations=equations,
+        limit=1,
+    )
+
+    transactions_builder = DiscoverQueryBuilder(
+        Dataset.Transactions,
+        params,
+        snuba_params=snuba_dataclass,
+        query=query,
+        selected_columns=selected_columns,
+        equations=equations,
+        limit=1,
+    )
+
+    dataset_inferred_from_query = dataset_split_decision_inferred_from_query(
+        errors_builder, transactions_builder
+    )
+
+    if dataset_inferred_from_query is not None:
+        if dry_run:
+            logger.info("Split decision for %s: %s", saved_query.id, dataset_inferred_from_query)
+        else:
+            save_split_decision_for_query(
+                saved_query,
+                dataset_inferred_from_query,
+                DatasetSourcesTypes.INFERRED.value,
+            )
+        return dataset_inferred_from_query, False
+
+    has_errors = False
+    try:
+        error_results = errors_builder.process_results(
+            errors_builder.run_query(
+                "backfill_discover_dataset", query_source=QuerySource.SENTRY_BACKEND
+            )
+        )
+        has_errors = len(error_results["data"]) > 0
+    except (snuba.QueryIllegalTypeOfArgument, snuba.UnqualifiedQueryError):
+        pass
+
+    if has_errors:
+        if dry_run:
+            logger.info(
+                "Split decision for %s: %s", saved_query.id, DiscoverSavedQueryTypes.ERROR_EVENTS
+            )
+        else:
+            save_split_decision_for_query(
+                saved_query,
+                DiscoverSavedQueryTypes.ERROR_EVENTS,
+                DatasetSourcesTypes.INFERRED.value,
+            )
+        return DiscoverSavedQueryTypes.ERROR_EVENTS, True
+
+    has_transactions = False
+    try:
+        transaction_results = transactions_builder.process_results(
+            transactions_builder.run_query(
+                "backfill_discover_dataset", query_source=QuerySource.SENTRY_BACKEND
+            )
+        )
+        has_transactions = len(transaction_results["data"]) > 0
+    except (snuba.QueryIllegalTypeOfArgument, snuba.UnqualifiedQueryError):
+        pass
+
+    if has_transactions:
+        if dry_run:
+            logger.info(
+                "Split decision for %s: %s",
+                saved_query.id,
+                DiscoverSavedQueryTypes.TRANSACTION_LIKE,
+            )
+        else:
+            save_split_decision_for_query(
+                saved_query,
+                DiscoverSavedQueryTypes.TRANSACTION_LIKE,
+                DatasetSourcesTypes.INFERRED.value,
+            )
+
+        return DiscoverSavedQueryTypes.TRANSACTION_LIKE, True
+
+    if dry_run:
+        logger.info(
+            "Split decision for %s: %s", saved_query.id, DiscoverSavedQueryTypes.TRANSACTION_LIKE
```

Comment:
```suggestion
            "Split decision for %s: %s", saved_query.id, DiscoverSavedQueryTypes.ERROR_EVENTS
```
Should we also record that the decision was forced?